File: src/com/sterndu/network/balancer/ConnectionEstablisher.java
package com.sterndu.network.balancer;

import java.io.*;
import java.net.*;
import java.util.*;
import com.sterndu.util.interfaces.*;

public class ConnectionEstablisher {

	public interface Connection extends Closeable {
		Socket s = new Socket();

		void bind(InetAddress address, Integer port) throws IOException;

		void bind(SocketAddress sa) throws IOException;

		void bind(String address, Integer port) throws IOException;

		void connect(InetAddress address, Integer port) throws IOException;

		void connect(InetAddress address, Integer port, int timeout) throws IOException;

		void connect(SocketAddress sa) throws IOException;

		void connect(SocketAddress sa, int timeout) throws IOException;

		void connect(String address, Integer port) throws IOException;

		void connect(String address, Integer port, int timeout) throws IOException;
		InetAddress getAddress();

		InputStream getInputStream() throws IOException;

		InetAddress getLocalAddress();

		int getLocalPort();

		OutputStream getOutputStream() throws IOException;

		int getPort();

		boolean isBound();

		boolean isClosed();

		boolean isConnected();
	}

	public interface Host extends Closeable {

		Connection acceptConnection() throws IOException;

		void bind(InetAddress address, int port) throws IOException;

		void bind(SocketAddress sa) throws IOException;

		void bind(String address, int port) throws IOException;

		InetAddress getInetAddress();

		int getLocalPort();

		SocketAddress getLocalSocketAddress();

		boolean isBound();

		boolean isClosed();

	}

	public static class Impl_Connection extends Socket implements Connection {
		public Impl_Connection() {

		}

		public Impl_Connection(InetAddress address, Integer port) throws IOException {
			super(address, port);
		}

		public Impl_Connection(InetSocketAddress sa) throws UnknownHostException, IOException {
			super(sa.getHostString(), sa.getPort());
		}

		public Impl_Connection(Socket s) throws IOException {
			super.bind(s.getLocalSocketAddress());
		}

		public Impl_Connection(String address, Integer port) throws UnknownHostException, IOException {
			super(address, port);
		}

		@Override
		public void bind(InetAddress address, Integer port) throws IOException {
			super.bind(new InetSocketAddress(address, port));
		}

		@Override
		public void bind(SocketAddress sa) throws IOException {
			super.bind(sa);
		}

		@Override
		public void bind(String address, Integer port) throws IOException {
			super.bind(new InetSocketAddress(address, port));
		}

		@Override
		public void connect(InetAddress address, Integer port) throws IOException {
			super.connect(new InetSocketAddress(address, port));
		}

		@Override
		public void connect(InetAddress address, Integer port, int timeout) throws IOException {
			super.connect(new InetSocketAddress(address, port), timeout);
		}

		@Override
		public void connect(SocketAddress sa) throws IOException {
			super.connect(sa);
		}

		@Override
		public void connect(SocketAddress sa, int timeout) throws IOException {
			super.connect(sa, timeout);
		}

		@Override
		public void connect(String address, Integer port) throws IOException {
			super.connect(new InetSocketAddress(address, port));
		}

		@Override
		public void connect(String address, Integer port, int timeout) throws IOException {
			super.connect(new InetSocketAddress(address, port), timeout);
		}

		@Override
		public InetAddress getAddress() { return super.getInetAddress(); }

		@Override
		public InetAddress getLocalAddress() { return super.getLocalAddress(); }
	}

	public static class Impl_Host extends ServerSocket implements Host {

		public Impl_Host() throws IOException {}

		public Impl_Host(int port) throws IOException {
			super(port);


		}

		public Impl_Host(int port, int backlog) throws IOException {
			super(port, backlog);
		}

		public Impl_Host(int port, int backlog, InetAddress bindAddr) throws IOException {
			super(port, backlog, bindAddr);
		}

		@Override
		public Connection acceptConnection() throws IOException {
			Impl_Connection c = new Impl_Connection();
			implAccept(c);
			return c;
		}

		@Override
		public void bind(InetAddress address, int port) throws IOException {
			super.bind(new InetSocketAddress(address, port));
		}

		@Override
		public void bind(String address, int port) throws IOException {
			super.bind(new InetSocketAddress(address, port));
		}

	}

	private static List<ThrowingBiFunction<String, Integer, Connection, IOException>> gateways = new ArrayList<>();

	private static List<ThrowingFunction<Integer, Host, IOException>> hosts = new ArrayList<>();

	static {
		addGateway(Impl_Connection::new);
		addHost(Impl_Host::new);
	}

	public static void addGateway(ThrowingBiFunction<String, Integer, Connection, IOException> gateway) {
		gateways.add(gateway);
	}

	public static void addGateway(ThrowingBiFunction<String, Integer, Connection, IOException> gateway,int index) {
		gateways.add(index,gateway);
	}

	public static void addHost(ThrowingFunction<Integer, Host, IOException> host) {
		hosts.add(host);
	}

	public static void addHost(ThrowingFunction<Integer, Host, IOException> host, int index) {
		hosts.add(index, host);
	}

	public static Connection establishConnection(String address, int port) throws IOException {
		for (ThrowingBiFunction<String, Integer, Connection, IOException> bi: gateways) try {
			return bi.apply(address, port);
		} catch (IOException e) {
		}
		throw new IOException("Could not connect:" + address + ":" + port);
	}

	public static Connection establishConnection(String address, int port, int index) throws IOException {
		try {
			return gateways.get(index).apply(address, port);
		} catch (IOException e) {
			throw new IOException("Could not connect:" + address + ":" + port);
		}
	}

	public static Host hostConnection(int port) throws IOException {
		for (ThrowingFunction<Integer, Host, IOException> bi: hosts) try {
			return bi.apply(port);
		} catch (IOException e) {
		}
		throw new IOException("Could not create:" + port);
	}

	public static Host hostConnection(int port, int index) throws IOException {
		try {
			return hosts.get(index).apply(port);
		} catch (IOException e) {
			throw new IOException("Could not create:" + port);
		}
	}

}

File: src/com/sterndu/network/balancer/ConnectionEstablisher.kt
@file:JvmName("ConnectionEstablisher")
package com.sterndu.network.balancer

import java.io.Closeable
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.*

object ConnectionEstablisher {
	private val gateways: MutableList<(String, Int) -> Connection> = ArrayList()
	private val hosts: MutableList<(Int) -> Host> = ArrayList()

	init {
		addGateway { address: String, port: Int ->
			ImplConnection(
				address, port
			)
		}
		addHost { port: Int -> ImplHost(port) }
	}

	fun addGateway(gateway: (String, Int) -> Connection) {
		gateways.add(gateway)
	}

	fun addGateway(gateway: (String, Int) -> Connection, index: Int) {
		gateways.add(index, gateway)
	}

	fun addHost(host: (Int) -> Host) {
		hosts.add(host)
	}

	fun addHost(host: (Int) -> Host, index: Int) {
		hosts.add(index, host)
	}

	@Throws(IOException::class)
	fun establishConnection(address: String, port: Int): Connection? {
		for (bi in gateways) try {
			return bi(address, port)
		} catch (e: IOException) {
			throw IOException("Could not connect:$address:$port")
		}
		return null
	}

	@Throws(IOException::class)
	fun establishConnection(address: String, port: Int, index: Int): Connection {
		return try {
			gateways[index](address, port)
		} catch (e: IOException) {
			throw IOException("Could not connect:$address:$port")
		}
	}

	@Throws(IOException::class)
	fun hostConnection(port: Int): Host? {
		for (bi in hosts) try {
			return bi(port)
		} catch (e: IOException) {
			throw IOException("Could not create:$port")
		}
		return null
	}

	@Throws(IOException::class)
	fun hostConnection(port: Int, index: Int): Host {
		return try {
			hosts[index](port)
		} catch (e: IOException) {
			throw IOException("Could not create:$port")
		}
	}

	interface Connection : Closeable {
		@Throws(IOException::class)
		fun bind(address: InetAddress, port: Int)

		@Throws(IOException::class)
		fun bind(sa: SocketAddress)

		@Throws(IOException::class)
		fun bind(address: String, port: Int)

		@Throws(IOException::class)
		fun connect(address: InetAddress, port: Int)

		@Throws(IOException::class)
		fun connect(address: InetAddress, port: Int, timeout: Int)

		@Throws(IOException::class)
		fun connect(sa: SocketAddress)

		@Throws(IOException::class)
		fun connect(sa: SocketAddress, timeout: Int)

		@Throws(IOException::class)
		fun connect(address: String, port: Int)

		@Throws(IOException::class)
		fun connect(address: String, port: Int, timeout: Int)
		fun getAddress(): InetAddress?

		@Throws(IOException::class)
		fun getInputStream(): InputStream?
		fun  getLocalAddress(): InetAddress
		fun  getLocalPort(): Int

		@Throws(IOException::class)
		fun getOutputStream(): OutputStream?
		fun  getPort(): Int
		fun  isBound(): Boolean
		fun  isClosed(): Boolean
		fun  isConnected(): Boolean

		companion object {
			val s = Socket()
		}
	}

	interface Host : Closeable {
		@Throws(IOException::class)
		fun acceptConnection(): Connection

		@Throws(IOException::class)
		fun bind(address: InetAddress, port: Int)

		@Throws(IOException::class)
		fun bind(sa: SocketAddress)

		@Throws(IOException::class)
		fun bind(address: String, port: Int)
		fun getInetAddress(): InetAddress?
		fun getLocalPort(): Int
		fun getLocalSocketAddress(): SocketAddress?
		fun isBound(): Boolean
		fun isClosed(): Boolean
	}

	class ImplConnection : Socket, Connection {
		constructor()
		constructor(address: InetAddress, port: Int) : super(address, port)
		constructor(sa: InetSocketAddress) : super(sa.hostString, sa.port)
		constructor(s: Socket) {
			super.bind(s.localSocketAddress)
		}

		constructor(address: String, port: Int) : super(address, port)

		@Throws(IOException::class)
		override fun bind(address: InetAddress, port: Int) {
			super.bind(InetSocketAddress(address, port))
		}

		@Throws(IOException::class)
		override fun bind(sa: SocketAddress) {
			super.bind(sa)
		}

		@Throws(IOException::class)
		override fun bind(address: String, port: Int) {
			super.bind(InetSocketAddress(address, port))
		}

		@Throws(IOException::class)
		override fun connect(address: InetAddress, port: Int) {
			super.connect(InetSocketAddress(address, port))
		}

		@Throws(IOException::class)
		override fun connect(address: InetAddress, port: Int, timeout: Int) {
			super.connect(InetSocketAddress(address, port), timeout)
		}

		@Throws(IOException::class)
		override fun connect(sa: SocketAddress) {
			super.connect(sa)
		}

		@Throws(IOException::class)
		override fun connect(sa: SocketAddress, timeout: Int) {
			super.connect(sa, timeout)
		}

		@Throws(IOException::class)
		override fun connect(address: String, port: Int) {
			super.connect(InetSocketAddress(address, port))
		}

		@Throws(IOException::class)
		override fun connect(address: String, port: Int, timeout: Int) {
			super.connect(InetSocketAddress(address, port), timeout)
		}

		override fun getAddress(): InetAddress? = super.getInetAddress()

		override fun getInputStream(): InputStream {
			return super.getInputStream()
		}

		override fun getLocalAddress(): InetAddress = super.getLocalAddress()
		override fun getLocalPort(): Int = super.getLocalPort()
		override fun getOutputStream(): OutputStream? {
			return super.getOutputStream()
		}
		override fun getPort(): Int = super.getPort()
		override fun isBound(): Boolean = super.isBound()
		override fun isClosed(): Boolean = super.isClosed()
		override fun isConnected(): Boolean = super.isConnected()

	}

	class ImplHost : ServerSocket, Host {
		constructor()
		constructor(port: Int) : super(port)
		constructor(port: Int, backlog: Int) : super(port, backlog)
		constructor(port: Int, backlog: Int, bindAddr: InetAddress) : super(port, backlog, bindAddr)

		@Throws(IOException::class)
		override fun acceptConnection(): Connection {
			val c = ImplConnection()
			implAccept(c)
			return c
		}

		@Throws(IOException::class)
		override fun bind(address: InetAddress, port: Int) {
			super.bind(InetSocketAddress(address, port))
		}

		@Throws(IOException::class)
		override fun bind(address: String, port: Int) {
			super.bind(InetSocketAddress(address, port))
		}

		override fun getInetAddress(): InetAddress? = super.getInetAddress()
		override fun getLocalPort(): Int = super.getLocalPort()
		override fun getLocalSocketAddress(): SocketAddress? = super.getLocalSocketAddress()
		override fun isBound(): Boolean = super.isBound()
		override fun isClosed(): Boolean = super.isClosed()
	}
}
